File: aliasing/aliasing.py
from dataclasses import dataclass

@dataclass
class NumberList:
    numbers: list[int]

    def take_all(self, other_list):
        self.numbers += other_list.numbers
        other_list.numbers = []

def test_take_all():
    list_a = NumberList([1, 2, 3])
    list_b = NumberList([4, 5, 6])
    list_a.take_all(list_b)
    assert list_a.numbers == [1, 2, 3, 4, 5, 6]
    assert list_b.numbers == []

    # OOPS: What does this do?
    list_a.take_all(list_a)
    assert list_a.numbers == [1, 2, 3, 4, 5, 6]

File: aliasing/aliasing.rs
pub struct NumberList {
    numbers: Vec<i32>,
}

impl NumberList {
    pub fn take_all(&mut self, other: &mut NumberList) {
        self.numbers.extend(&other.numbers);
        other.numbers.clear();
    }
}

fn main() {
    let mut list_a = NumberList {
        numbers: vec![1, 2, 3],
    };
    let mut list_b = NumberList {
        numbers: vec![4, 5, 6],
    };
    list_a.take_all(&mut list_b);
    assert_eq!(list_a.numbers, [1, 2, 3, 4, 5, 6]);
    assert_eq!(list_b.numbers, []);

    // OOPS: What does this do?
    list_a.take_all(&mut list_a);
    assert_eq!(list_a.numbers, [1, 2, 3, 4, 5, 6]);
}
